Update square-root description.md (#2495)
exercism/problem-specifications: exercises/square-root/instructions.md

# Instructions

> Your task is to implement a calculation of the square root of a given number.

Reviewer: ```suggestion
Your task is to calculate the square root of a given number.
```

- Try to avoid using the pre-existing math libraries of your language.

> - As input you’ll be given a positive whole number, i.e. 1, 2, 3, 4…

Reviewer 2: ```suggestion
- As input you'll be given a positive whole number, i.e. 1, 2, 3, 4…
```

> - You also only need to output a positive whole number.

Reviewer 3: Is the point here that we're getting the nearest whole number? (ie approximating/rounding?)

Author: Yes.  This was chosen to narrow the scope of the exercise and hopefully make it both easier and more interesting.  It also opens the possibility of more alternative implementations such as linear and binary search.

> - You also only need to output a positive whole number.

Reviewer: ```suggestion
- You are only required to handle cases where the output is a positive whole number.
```

Author: The suggested change is misleading because it seems to preclude approximating the output to a positive whole number even when it is not the exact result, which most implementations of integer square root will do.  It biases the student to implementations that only handle perfect squares, which overly narrows the exercise.

Reviewer 4: I think this is a good suggestion, we can see what the requirements are from the tests.  What is allowed will pass the tests, even (and partly because of) the tests not being exhaustive.

The tests guide what the requirement is, and potentially directly what is not allowed.

The reading can be a bit ambiguous, but may  not contradict.

**What about leaving that line out?**  We can see what the tests require, and the tests are going to communicate, and we also encourage the students to read them.

Also, "output" does not happen in the sense of "to  a screen or file" but what results is a positive whole number.

The tests are also a source of bias, though, we may not be able to avoid that.

Author: OK, I've replaced "output" with "return".

Reviewer 2: I second IsaacG's suggestion. It emphasizes that we are limited to inputs that yield a specific type of result. The current change just focuses on what the result is. The difference is subtle but important.

> Examples of some approaches you could consider are:

Reviewer: ```suggestion
Here are some approaches you could use.
```

> - Linear or binary search for a number that gives the input number when squared

Reviewer: These sentences should have a period at the end

- Successive approximation using Newton's or Heron's method.
- Calculating one digit at a time or one bit at a time.

> You can check out the Wikipedia pages on [integer square root][integer-square-root] and [methods of computing square roots][computing-square-roots] to help with choosing a method of calculation.

Reviewer 2: ```suggestion
You can check out the Wikipedia pages on [methods of computing square roots][computing-square-roots] to help with choosing a method of calculation.
```

> [integer-square-root]: https://en.wikipedia.org/wiki/Integer_square_root

Reviewer 2: ```suggestion
```

The integer square root differs from the usual square root, and the existing wiki link already provides enough information. It might lead students down an unintended path and cause extra confusion.

Author: I thought the idea was to have students specifically implement integer square root, not the full floating point square root, as a way of narrowing down the scope of the task.  The backstory was chosen partly to support this.

Author: Additional reasons why I intentionally want to explicitly require the integer square root in the problem specification:

* Algorithms that compute the non-integer square root run forever on inputs that are not a perfect square.  The integer square root algorithms do not run forever.
* The existing submissions, at least in the Python track, have taken their cues from the tests expecting integer results and the students have typically truncated the results to an integer even when using the built in operators, for example many submissions are of the form int(n^0.5).
* The integer square root has a wider range of alternative implementations as suggested in these new instructions, which I think makes the exercise more interesting.

Therefore if any Wikipedia link is to be removed I would prefer it be the one to the "Computing square roots" page, not the "Integer square roots" page.  The inverse of your requested change.

Reviewer 5: I'm with @xanni on this one. Judging from the original implementation, the author meant this exercise as an exercise on perfect squares.
I also thin that both links can stay. Not that many people will bother following up two links before solving the exercise. If they get stuck, they loo to community solutions instead.

Reviewer 4: Not to mention the risk of them doing their own research and being exposed to things.

Author: Wikipedia links removed.

# Instructions

Your task is to calculate the square root of a given number.

- Try to avoid using the pre-existing math libraries of your language.
- As input you’ll be given a positive whole number, i.e. 1, 2, 3, 4…
- You also only need to output a positive whole number.

> Here are some approaches you could use.

Reviewer 4: ```suggestion
Some potential approaches:
```

Since what follows is a bullet list.
